Description: Cache the result of ck-history so greeter shows user list immediately after login
Bug: http://bugzilla.gnome.org/show_bug.cgi?id=594344
Bug-Ubuntu: https://bugs.edge.launchpad.net/ubuntu/karmic/+source/gdm/+bug/400863

Index: gdm-2.30.2/gui/simple-greeter/Makefile.am
===================================================================
--- gdm-2.30.2.orig/gui/simple-greeter/Makefile.am	2010-04-27 03:08:42.000000000 +1000
+++ gdm-2.30.2/gui/simple-greeter/Makefile.am	2010-06-22 13:11:02.694148945 +1000
@@ -10,6 +10,7 @@
 	-DDMCONFDIR=\""$(dmconfdir)"\"			\
 	-DGDMCONFDIR=\"$(gdmconfdir)\"                  \
 	-DDATADIR=\""$(datadir)"\"		 	\
+	-DCACHEDIR=\""$(GDM_WORKING_DIR)/.cache"\"	 	\
 	-DSYSCONFDIR=\""$(sysconfdir)"\"		\
 	-DLIBLOCALEDIR=\""$(prefix)/lib/locale"\"	\
 	-DGNOMELOCALEDIR=\""$(datadir)/locale"\" 	\
Index: gdm-2.30.2/gui/simple-greeter/gdm-user-manager.c
===================================================================
--- gdm-2.30.2.orig/gui/simple-greeter/gdm-user-manager.c	2010-06-22 13:10:23.244157884 +1000
+++ gdm-2.30.2/gui/simple-greeter/gdm-user-manager.c	2010-06-22 13:11:02.694148945 +1000
@@ -78,6 +78,8 @@
 #define DEFAULT_GLOBAL_FACE_DIR DATADIR "/faces"
 #define DEFAULT_USER_ICON       "stock_person"
 
+#define LOGIN_CACHE_FILE CACHEDIR "/login_frequency.cache"
+
 struct GdmUserManagerPrivate
 {
         GHashTable            *users;
@@ -98,6 +100,8 @@
         guint                  ck_history_id;
 
         guint8                 users_dirty : 1;
+        guint8                 loaded_cache : 1;
+        guint8                 loading_users : 1;    
 };
 
 enum {
@@ -1292,7 +1296,30 @@
         }
 
         if (done) {
-                g_signal_emit (G_OBJECT (manager), signals[USERS_LOADED], 0);
+                FILE *fp;
+
+                /* Cache login counts */
+                fp = fopen (LOGIN_CACHE_FILE, "w");
+                if (fp != NULL) {
+                        GHashTableIter iter;
+                        gpointer value;
+
+                        g_hash_table_iter_init (&iter, manager->priv->users);
+                        while (g_hash_table_iter_next (&iter, NULL, &value)) {
+                                GdmUser *user = (GdmUser *) value;
+                                fprintf (fp, "%s %lu\n",
+                                         gdm_user_get_user_name (user),
+                                         gdm_user_get_login_frequency (user));
+                        }
+                        fclose (fp);
+                }
+                else
+                        g_warning ("Unable to write to login cache file: %s", LOGIN_CACHE_FILE);
+
+                if (manager->priv->loading_users) {
+                        g_signal_emit (G_OBJECT (manager), signals[USERS_LOADED], 0);
+                        manager->priv->loading_users = FALSE;
+                }
 
                 manager->priv->ck_history_id = 0;
                 return FALSE;
@@ -1531,8 +1558,35 @@
 }
 
 static void
+load_login_frequency_cache (GdmUserManager *manager)
+{
+        GIOChannel *channel;
+        gchar *line;
+
+        channel = g_io_channel_new_file (LOGIN_CACHE_FILE, "r", NULL);
+        if (channel == NULL)
+                return;
+
+        while (g_io_channel_read_line (channel, &line, NULL, NULL, NULL) == G_IO_STATUS_NORMAL) {
+                process_ck_history_line (manager, line);
+                g_free (line);
+        }
+    
+        g_io_channel_close (channel);
+    
+        if (manager->priv->loading_users) {
+                g_signal_emit (G_OBJECT (manager), signals[USERS_LOADED], 0);
+                manager->priv->loading_users = FALSE;
+        }    
+}
+
+static void
 reload_users (GdmUserManager *manager)
 {
+        if (!manager->priv->loaded_cache) {
+                load_login_frequency_cache (manager);
+                manager->priv->loaded_cache = TRUE;
+        }
         reload_ck_history (manager);
         reload_passwd (manager);
 }
@@ -1554,6 +1608,7 @@
         }
 
         g_signal_emit (G_OBJECT (manager), signals[LOADING_USERS], 0);
+        manager->priv->loading_users = TRUE;
         manager->priv->reload_id = g_idle_add ((GSourceFunc)reload_users_timeout, manager);
 }
 
